File: ws-go/tests/integration_test.go
// Copyright 2026 TEAM RelicFrog
// SPDX-License-Identifier: Apache-2.0
//
// integration_test.go — Integration tests for the primes package.
//
// These tests exercise the public API end-to-end, mirroring the
// integration tests in the Rust ws-rust workspace.

package primes_test

import (
	"errors"
	"testing"

	primes "github.com/RelicFrog/ops-devbox-examples/ws-go/src"
)

// ---------------------------------------------------------------------------
// Round-trip: NthPrime is consistent with SieveOfEratosthenes
// ---------------------------------------------------------------------------

func TestNthPrimeConsistentWithSieveFirst50(t *testing.T) {
	t.Parallel()

	sieve, err := primes.SieveOfEratosthenes(230) // 230 > 50th prime (229)
	if err != nil {
		t.Fatal(err)
	}

	for i, expected := range sieve {
		idx := uint64(i + 1)

		p, err := primes.NthPrime(idx)
		if err != nil {
			t.Fatalf("NthPrime(%d): unexpected error: %v", idx, err)
		}

		if p != expected {
			t.Errorf("NthPrime(%d) = %d, want %d", idx, p, expected)
		}
	}
}

// ---------------------------------------------------------------------------
// PrimesInRange covers full first 100 primes
// ---------------------------------------------------------------------------

func TestPrimesInRangeFirstHundredMatchesSieve(t *testing.T) {
	t.Parallel()

	sieve, err := primes.SieveOfEratosthenes(541) // 541 == 100th prime
	if err != nil {
		t.Fatal(err)
	}

	rang, err := primes.PrimesInRange(2, 541)
	if err != nil {
		t.Fatal(err)
	}

	if len(rang) != 100 {
		t.Fatalf("PrimesInRange(2,541): got %d primes, want 100", len(rang))
	}

	for i := range sieve {
		if sieve[i] != rang[i] {
			t.Errorf("index %d: sieve=%d range=%d", i, sieve[i], rang[i])
		}
	}
}

// ---------------------------------------------------------------------------
// IsPrime consistent with sieve for 0..200
// ---------------------------------------------------------------------------

func TestIsPrimeConsistentWithSieveUpTo200(t *testing.T) {
	t.Parallel()

	sieve, err := primes.SieveOfEratosthenes(200)
	if err != nil {
		t.Fatal(err)
	}

	sieveSet := make(map[uint64]bool, len(sieve))
	for _, p := range sieve {
		sieveSet[p] = true
	}

	for n := uint64(0); n <= 200; n++ {
		got := primes.IsPrime(n)
		want := sieveSet[n]

		if got != want {
			t.Errorf("IsPrime(%d) = %v, sieve says %v", n, got, want)
		}
	}
}

// ---------------------------------------------------------------------------
// Error propagation
// ---------------------------------------------------------------------------

func TestSieveLimitZeroPropagatesError(t *testing.T) {
	t.Parallel()

	_, err := primes.SieveOfEratosthenes(0)
	if !errors.Is(err, primes.ErrLimitTooSmall) {
		t.Errorf("expected ErrLimitTooSmall, got %v", err)
	}
}

func TestPrimesInRangeInvertedPropagatesError(t *testing.T) {
	t.Parallel()

	_, err := primes.PrimesInRange(50, 10)

	var rangeErr *primes.ErrInvalidRange
	if !errors.As(err, &rangeErr) {
		t.Errorf("expected *ErrInvalidRange, got %T: %v", err, err)
	}
}

func TestNthPrimeZeroPropagatesError(t *testing.T) {
	t.Parallel()

	_, err := primes.NthPrime(0)
	if !errors.Is(err, primes.ErrInvalidNthIndex) {
		t.Errorf("expected ErrInvalidNthIndex, got %v", err)
	}
}

// ---------------------------------------------------------------------------
// Known large primes / composites
// ---------------------------------------------------------------------------

func TestKnownLargePrimesAreDetected(t *testing.T) {
	t.Parallel()

	for _, p := range []uint64{7919, 104729, 999_983} {
		if !primes.IsPrime(p) {
			t.Errorf("IsPrime(%d) = false, want true", p)
		}
	}
}

func TestKnownLargeCompositesAreRejected(t *testing.T) {
	t.Parallel()

	for _, c := range []uint64{7920, 104728, 999_999} {
		if primes.IsPrime(c) {
			t.Errorf("IsPrime(%d) = true, want false", c)
		}
	}
}

File: ws-lua/tests/integration_test.lua
-- Copyright 2026 TEAM RelicFrog
-- SPDX-License-Identifier: Apache-2.0
--
-- tests/integration_test.lua — Integration tests for primes module.

local primes = require("primes")

local pass = 0
local fail = 0

local function ok(desc, cond)
  if cond then
    pass = pass + 1
    io.write(string.format("  ok  %d — %s\n", pass + fail, desc))
  else
    fail = fail + 1
    io.write(string.format("  FAIL %d — %s\n", pass + fail, desc))
  end
end

local function eq_list(a, b)
  if #a ~= #b then return false end
  for i = 1, #a do
    if a[i] ~= b[i] then return false end
  end
  return true
end

-- Round-trip: nth_prime consistent with sieve (first 50)
local sieve230 = primes.sieve_of_eratosthenes(230)
local all_match = true
for i, expected in ipairs(sieve230) do
  if primes.nth_prime(i) ~= expected then all_match = false; break end
end
ok("nth_prime consistent with sieve (first " .. #sieve230 .. " primes)", all_match)

-- primesInRange first 100 matches sieve
local sieve541  = primes.sieve_of_eratosthenes(541)
local range541  = primes.primes_in_range(2, 541)
ok("primesInRange(2,541) length == 100", #range541 == 100)
ok("primesInRange(2,541) matches sieve", eq_list(range541, sieve541))

-- isPrime consistent with sieve for 0..200
local sieve200 = primes.sieve_of_eratosthenes(200)
local sieve_set = {}
for _, p in ipairs(sieve200) do sieve_set[p] = true end
local consistent = true
for n = 0, 200 do
  if primes.is_prime(n) ~= (sieve_set[n] == true) then
    consistent = false; break
  end
end
ok("is_prime consistent with sieve for 0..200", consistent)

-- Error propagation
local ok1, _ = pcall(primes.sieve_of_eratosthenes, 0)
ok("sieve(0) raises",               not ok1)
local ok2, _ = pcall(primes.primes_in_range, 50, 10)
ok("primes_in_range(50,10) raises", not ok2)
local ok3, _ = pcall(primes.nth_prime, 0)
ok("nth_prime(0) raises",           not ok3)

-- Known large primes / composites
for _, p in ipairs({7919, 104729, 999983}) do
  ok("is_prime(" .. p .. ") == true",  primes.is_prime(p))
end
for _, c in ipairs({7920, 104728, 999999}) do
  ok("is_prime(" .. c .. ") == false", not primes.is_prime(c))
end

print(string.format("\n%d integration tests run: %d passed, %d failed", pass + fail, pass, fail))
if fail > 0 then os.exit(1) end
